core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala

```diff
@@ -199,41 +242,137 @@ private[spark] class HeartbeatReceiver(sc: SparkContext, clock: Clock)
+          case _: LocalSchedulerBackend =>
+
+          case other => throw new UnsupportedOperationException(
+            s"Unknown scheduler backend: ${other.getClass}")
+        }
+      }
+    })
+  }
+
+  private def removeExecutorFromWaitingList(executorId: String): Unit = {
+    val isStandalone = sc.schedulerBackend.isInstanceOf[StandaloneSchedulerBackend]
+    if (checkWorkerLastHeartbeat && isStandalone) {
+      waitingList.remove(executorId)
+    }
+  }
+
   private def expireDeadHosts(): Unit = {
+  /**
+   * [SPARK-39984]
+   * Originally, the driver’s HeartbeatReceiver will expire an executor if it does not receive any
+   * heartbeat from the executor for 120 seconds. However, 120 seconds is too long, but we will face
+   * other challenges when we try to lower the timeout threshold. To elaborate, when an executor is
+   * performing full GC, it cannot send/reply any message. Next paragraphs describe the solution to
+   * detect network disconnection between driver and executor in a short time.
+   *
+   * An executor is running on a worker but in different JVMs, and a driver is running on a master
+   * but in different JVMs. Hence, the network connection between driver/executor and master/worker
+   * is the same. Because executor and worker are running on different JVMs, worker can still send
+   * heartbeat to master when executor performs GC.
+   *
+   * For new Heartbeat Receiver, if driver does not receive any heartbeat from the executor for
+   * `executorTimeoutMs` (default: 60s) seconds, HeartbeatReceiver will send a request to master to
+   * ask for the latest heartbeat from the worker which the executor runs on `workerLastHeartbeat`.
+   * HeartbeatReceiver can determine whether the heartbeat loss is caused by network issues or other
+   * issues (e.g. GC). If the heartbeat loss is not caused by network issues, the HeartbeatReceiver
+   * will put the executor into a waitingList rather than expiring it immediately.
+   *
+   * [Note]: Definition of `network issues`
+   * Here, the definition `network issues` is the issues that related to network directly. If the
+   * network is connected, the issues do not included in `network issues`. For example, an
+   * executor's JVM is closed by a problematic task, so the JVM will notify driver that the socket
+   * is closed. If the network is connected, driver will receive the notification and trigger the
+   * function `onDisconnected`. This issue is not a `network issue` because the network is
+   * connected.
+   *
+   * [Warning 1]
+   * Worker will send heartbeats to Master every (conf.get(WORKER_TIMEOUT) * 1000 / 4) milliseconds.
+   * Check deploy/worker/Worker.scala for more details. This new mechanism design is based on the
+   * assumption: `waitingListTimeout` > (conf.get(WORKER_TIMEOUT) * 1000 / 4).
+   *
+   * [Warning 2]
+   * Not every deployment method schedules driver on master.
+   */
```

Comment:
Let us make this comment more concise, and what the state would be after this PR is merged.

Something like
```suggestion
  /**
   * [SPARK-39984]
   * The driver’s HeartbeatReceiver will expire an executor if it does not receive any
   * heartbeat from the executor for 120 seconds. However, lowering from 120 seconds has 
   * other challenges. For example: when executor is performing full GC it cannot send/reply any message.
   *
   * For standalone deployments:
   * If driver does not receive any heartbeat from the executor for
   * `executorTimeoutMs` (default: 60s) seconds, HeartbeatReceiver will send a request to master to
   * ask for the latest heartbeat from the worker which the executor runs on.
   * HeartbeatReceiver can determine whether the heartbeat loss is caused by network issues or other
   * issues (e.g. GC). If the heartbeat loss is not caused by network issues, the HeartbeatReceiver
   * will put the executor into a waitingList rather than expiring it immediately.
   */
```

Also, let us move this to the configuration of `HEARTBEAT_WAITINGLIST_TIMEOUT`:
```
   * assumption: `waitingListTimeout` > (conf.get(WORKER_TIMEOUT) * 1000 / 4).
```

If this is a strong requirement, add as warning when fetching the the config above ?

Reply:
(Note: `waitingListTimeout` is renamed to `expiryCandidatesTimeout`)
```
   * assumption: waitingListTimeout > (conf.get(WORKER_TIMEOUT) * 1000 / 4).
```

This is not a strong requirement, but it will make more sense if we fulfill this requirement. If we fulfill requirement, we can know whether master lost any heartbeat from the worker or not.

I can add it as a warning, but I am not sure what is a "warning". Do you mean `logWarning`? Thank you!

Reply:
Yes, log a warning

Reply:
Added it as a warning, and simplified comments.